### ISSUE TITLE
Are NBVS and TS equivalent to a special FC layer?

### ISSUE THREAD

**Issue by ShellingFord221** [blindauth/labelshiftexperiments#3]
Hi @AvantiShri, long time no see! After two months, I have a new idea. Are NBVS and TS equivalent to a special FC layer, where the vector for each class decays to a scalar (i.e. temperature), and we fine-tune this layer only on the held-out set of training data? If so, does that mean that we can simply add one more FC layer after the final FC layer and may get better accuracy than NBVS and TS? (because they just do calibration by a scalar for each class, bias term remains the same)
Thank you very much!

**Comment by AvantiShri** [comment]
Hi Shelling, yes, TS, BCTS, NBVS and VS can all be implemented as a type of FC layer where certain weights are set to zero and other weights are constrained to be equal. In the case of TS, all the bias terms and off-diagonal elements in the weight matrix are set to zero, while on-diagonal elements are set to be equal to the temperature. BCTS is like TS, but bias terms are now included. VS is like BCTS, but the on-diagonal elements are allowed to vary independently of each other - and so forth.

"If so, does that mean that we can simply add one more FC layer after the final FC layer and may get better accuracy than NBVS and TS?" - using a full FC layer instead of TS/BCTS/NBVS/VS would be equivalent to "matrix scaling", which was discussed in the Guo et al. temperature scaling paper. Matrix scaling was generally found to overfit. Matrix scaling with something called ODIR regularization has been found to give decent performance, though: https://papers.nips.cc/paper/9397-beyond-temperature-scaling-obtaining-well-calibrated-multi-class-probabilities-with-dirichlet-calibration.pdf
